src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -223,6 +223,12 @@ void Cancel()
                     if (ipAddresses[i] != null)
                     {
                         sockets[i] = new Socket(ipAddresses[i].AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+#if FEATURE_TCPKEEPALIVE
+                        // enable keep-alive on socket
+                        sockets[i].SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
+                        sockets[i].SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, 1000);
+                        sockets[i].SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, 30);
```

Comment:
This is in minutes, right?

Reply:
Good question .. I am trying to look for documentation . I see here it mentions in milliseconds:
[https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc782936(v=ws.10)](https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc782936(v=ws.10))

This user also mentioned the options are milliseconds here:
[https://github.com/dotnet/corefx/issues/25040](https://github.com/dotnet/corefx/issues/25040)

Reply:
We need 30 minutes. 30 Millis doesn't make sense. I wonder how Karl came up with this value. Can you sync with him?
Native SNI uses 30 minutes

Reply:
According to these links:
1- https://msdn.microsoft.com/en-us/library/windows/desktop/dd877220%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396

2- http://man7.org/linux/man-pages/man7/tcp.7.html

The measurements for the `TcpKeepAliveInterval `and `TcpKeepAliveTime `in Windows and Linux OS are different. In Linux the `TCP_KEEPINTVL` and `TCP_KEEPIDLE ` is set in seconds, while in Windows these values are measured in milliseconds. 
Should we have different values based on different platform? 
@stephentoub @karelz  Do you know if this is something we can get some guidance from System.Net.sockets area owners?

Reply:
> Should we have different values based on different platform?
@stephentoub @karelz Do you know if this is something we can get some guidance from System.Net.sockets area owners?

No, otherwise it makes it really difficult to use.  Since these are already different settings on the different platforms and the Sockets class is trying to provide an abstraction over them, it should also be normalizing the units.  If it's not, that's a bug and we should fix it.

@AfsanehR, can you open an issue tracking that?  Thanks.

cc: @luigiberrettini

Reply:
As highlighted in #25040 `TCP_KEEPALIVE` / `TCP_KEEPIDLE` and `TCP_KEEPINTVL` are expressed in seconds while `SIO_KEEPALIVE_VALS` are [expressed in milliseconds](https://msdn.microsoft.com/en-us/library/windows/desktop/dd877220).

Since #29963 allows to set keep-alive retry count, time and interval on all platforms without too many abstractions, it is up to the developer the logic that allow to use the proper socket method depending on the operating system and its version.

Essentially:
```csharp
```

 if (isWin10V1703OrLater)
                Socket.SetSocketOption(SocketOptionLevel.Tcp, TcpKeepAliveRetryCount, keepAliveConfig.RetryCount);

            if (isBelowWin10V1709)
            {
                // Call WSAIoctl via IOControl
                Socket.IOControl(IOControlCode.KeepAliveValues, new IOControlKeepAliveValues(keepAliveConfig).ToByteArray(), null);
                return;
            }
            Socket.SetSocketOption(SocketOptionLevel.Tcp, TcpKeepAliveTime, keepAliveConfig.Time);
Socket.SetSocketOption(SocketOptionLevel.Tcp, TcpKeepAliveInterval, keepAliveConfig.Interval);


 TCP_KEEPALIVE) and TCP_KEEPINTVL are expressed in seconds, while SIO_KEEPALIVE_VALS expresses durations in milliseconds.
```suggestion
                        sockets[i].SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, 30);
```

Reply:
As highlighted in #25040 `TCP_KEEPALIVE` / `TCP_KEEPIDLE` and `TCP_KEEPINTVL` are expressed in seconds while `SIO_KEEPALIVE_VALS` are [expressed in milliseconds](https://msdn.microsoft.com/en-us/library/windows/desktop/dd877220).

Since #29963 allows to set keep-alive retry count, time and interval on all platforms, but without making the developer life easier, there is still the need to code the logic that allow to use the proper socket method depending on the operating system and its version:

```csharp
//  *** Retry count
// Linux, OS X, Windows >= 1703
socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveRetryCount, RetryCount);

// *** Time and interval
// Windows < 1709
socket.IOControl(IOControlCode.KeepAliveValues, byteArrayWithTimeAndIntervalMilliseconds, null);
// Linux, OS X, Windows >= 1709
socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, TimeSeconds);
socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, IntervalSeconds);
```

Reply:
> there is still the need to code the logic that allow to use the proper socket method depending on the operating system and its version:

For time and interval, why can't we just make this work from the .net layer? We'd just translate the SetSocketOption to the appropriate call on a lower version of Windows, and we'd ensure the. NET API is consistent on units.

Reply:
If a developer wants to set only one of them which value would you use for the other to pass the byte array to IOControl?
When a user tries to set both with two different SetSocketOption calls, the second call will override what has been set by the first call.

The only way I see is either keeps things as they are (every dev will have to write the logic) or create an API that expect a bool (enable/disable) and the three values to be set: this would allow performing individual SetSocketOption or other calls under the hood.

Reply:
@AfsanehR I want to correct my statement, the keep alive time should be 30 seconds. And the KeepAlive internal should be 1 second. 
Values from Windows code in milliseconds.
```
DWORD const KEEPALIVETIME = 30000;
DWORD const KEEPALIVEINTERVAL = 1000;

```